### ISSUE TITLE
gopass audit spams multiple password prompts/requests

### ISSUE THREAD

**Issue by iicurtis** [gopasspw/gopass#1017]
### Summary

`gopass audit` opens the password dialog requesting the pgp secret. However, as soon as the secret is entered, another dialog requesting the same thing opens. This happens repeatedly. Furthermore, as timeouts occur, the `gopass audit` is filled with 
```
gpg: decryption failed: No secret key
1 of 587 secrets checked                                                                      0.17%
gpg: decryption failed: No secret key
11 of 587 secrets checked                                                                     1.87%
gpg: decryption failed: No secret key
13 of 587 secrets checked                                                                     2.21%
gpg: decryption failed: No secret key
475 of 587 secrets checked                                                                   80.92%
gpg: decryption failed: No secret key
534 of 587 secrets checked                                                                   90.97%
gpg: decryption failed: No secret key
576 of 587 secrets checked                                                                   98.13%
gpg: decryption failed: No secret key
gpg: decryption failed: No secret key
579 of 587 secrets checked                                                                   98.64%
gpg: decryption failed: No secret key
gpg: decryption failed: No secret keygpg: decryption failed: No secret key
```

### Steps To Reproduce

Run `gopass audit`. 

### Expected behavior

Only one prompt for secrets should be required.

### Environment

- OS: Arch Linux 4.19.9-zen
- DE: KDE Plasma 5.14.4
- gopass version: 1.8.3
- Installation method: pacman community package

**Comment by dragon788** [comment]
@iicurtis what does your `~/.gnupg/gpg-agent.conf` and `~/.gnupg/gpg.conf` have in them for timeout values and `pinentry` program?

**Comment by nylocx** [comment]
That is strange I have the exact same effect during audit it askes for the PW multiple time and later reports that some secrets could not be decrypter. But if I run them manually I can view and decrypt them just fine.

Arch Linux
gopass 1.8.6
fish shell
/usr/bin/pinentry as pinentry binary
pinentry timeout 86400

~/.gnupg/gpg.conf does not exist

**Comment by c0dev0id** [comment]
Same here. Everything in gopass works, but audit asks for passwords. The same passwords can be decrypted normally in gopass.
`gopass 1.14.10 go1.19.2 openbsd amd64`

Maybe it would help to throttle the audit process? It feels like gpg-agent can't satisfy all requests properly...

**Comment by dominikschulz** [comment]
This is a strange issue. If using the GPG backend we should limit concurrency to 1 already, but maybe that's buggy. Or it's yet another issue.

@c0dev0id Could you maybe try if `gopass fsck` triggers that same issue?

https://github.com/gopasspw/gopass/pull/2498 is adding some more logging, just in case.

**Comment by c0dev0id** [comment]
No, gopass fsck runs fine and does not trigger any password prompts.

**Comment by c0dev0id** [comment]
I recorded stdout with gpg verbose enabled. This doesn't look very sequential to me: https://codevoid.de/0/p/gopass_stdout.txt

You can also see "pinentry launched" every once in a while

**Comment by c0dev0id** [comment]
Just fyi. Setting "audit.concurrency = 1" fixes the issue and runs stable.

On my system, 2,3,4 also runs stable. I start seeing PW prompts with 5.

**Comment by bdebyl** [comment]
> Just fyi. Setting "audit.concurrency = 1" fixes the issue and runs stable.
> 
> On my system, 2,3 also runs stable. I start seeing PW prompts with 4.

Same, for those with the issue in my `~/.config/gopass/config` I added it as follows:

```
[audit]
  concurrency = 1
```

**Comment by leogott** [comment]
I'm running into a very similar issue, running `gopass audit` not quite consistently fails to decrypt around 1% of secrets with a gpg missing key error.
```
gpg: Entschlüsselung fehlgeschlagen: Kein geheimer Schlüssel
❌ Decryption failed: exit status 2
```
In two successive runs (without the workaround) I first got errors on the first two secrets, and then only on number 47 .
(in both cases I had to enter my pin two or three times, which isn't too bad)

**I believe it's a race condition related to the gpg-agent. Spawning a new thread while the agent is busy starting pinentry or something like that... that just shows up on bulk commands more often.**
I noticed a pattern: *most of the time,* gopass audit shows a new gpg error right when a new pinentry opens (not always though. Sometimes there is quite a delay)

I tried the workaround `gopass config audit.concurrency 1`, and it seems to work without any extra pin-entries, but is sloooow.

Then I tried concurrency 16, which is _blazing fast_, but opens pin-entries almost back to back.
Curiously, I *didn't get any* gpg errors the first time I tried this, but in successive attempts I always got a few.

I also tried running audit in the repl, or after running `gpgconf --kill gpg-agent` and both in the repl after killing the agent.

Anecdotally, I get fewer errors on the repl. Curiously killing the agent doesn't ensure that the first secret gets decrypted successfully.

Mostly unrelated, but watching ~/.gnupg ... could it be that gopass audit never has more than 10 decryptions running simultaneously?

gopass 1.15.12 go1.22.1 linux amd64
gpg 2.2.27 - gitfs 2.43.2
gpg (GnuPG) 2.2.27
libgcrypt 1.9.4
